src/implicit.rs

```diff
@@ -137,19 +139,140 @@ pub fn resolve_weak(sequence: &IsolatingRunSequence, processing_classes: &mut [B
+
+    // Identify the bracket pairs in the current isolating run sequence according to BD16.
+    let bracket_pairs = identify_bracket_pairs(text, data_source, sequence, original_classes);
+
+    // For each bracket-pair element in the list of pairs of text positions
+    // Note: Rust ranges are interpreted as [start..end), be careful using `pair` directly
+    // for indexing as it will include the opening bracket pair but not the closing one
+    for pair in bracket_pairs {
+        #[cfg(feature = "std")]
+        debug_assert!(
+            pair.start < processing_classes.len(),
+            "identify_bracket_pairs returned a range that is out of bounds!"
+        );
+        #[cfg(feature = "std")]
+        debug_assert!(
+            pair.end < processing_classes.len(),
+            "identify_bracket_pairs returned a range that is out of bounds!"
+        );
+        let mut found_l = false;
+        let mut found_r = false;
+        let mut class_to_set = None;
+        // Inspect the bidirectional types of the characters enclosed within the bracket pair.
+        // Note: the algorithm wants us to inspect the types of the *enclosed* characters,
+        // not the brackets themselves, however since the brackets will never be L or R, we can
+        // just scan them as well and not worry about trying to skip them in the array (they may take
+        // up multiple indices in processing_classes if they're multibyte!).
+        //
+        // `pair` is [start, end) so we will end up processing the opening character but not the closing one.
+        //
+        // Note: Given that processing_classes has been modified in the previous runs, and resolve_weak
+        // modifies processing_classes inconsistently at non-character-boundaries,
+        // this and the later iteration will end up iterating over some obsolete classes.
+        // This is fine since all we care about is looking for strong
+        // classes, and strong_classes do not change in resolve_weak. The alternative is calling `.char_indices()`
+        // on the text (or checking `text.get(idx).is_some()`), which would be a way to avoid hitting these
+        // processing_classes of bytes not on character boundaries. This is both cleaner and likely to be faster
+        // (this is worth benchmarking, though!) so we'll stick with the current approach of iterating over processing_classes.
+        for &class in &processing_classes[pair.clone()] {
+            if class == BidiClass::R {
+                found_r = true;
+            } else if class == BidiClass::L {
+                found_l = true;
+            }
+
+            // if we have found a character with the class of the embedding direction
+            // we can bail early
+            if found_l && e_is_l || found_r {
+                break;
+            }
+        }
+        // If any strong type (either L or R) matching the embedding direction is found
+        if found_l && e_is_l || found_r {
```

Comment:
I think it would be easier to read and reason about the implementation if the booleans were `found_e` and `found_opposite_e`, this expression comes up twice and is uncomfortably asymmetric.